### PR TITLE
Module4

### PR DESCRIPTION
**Pull request by ralucaromascu**

### CODE REVIEW

File: module4/ex4_loglistclass/ex4_loglistclass_test.py [hunk 1]
```diff
@@ -6,15 +6,15 @@
 
 def test_dicts_returns_list_of_dicts():
     ld = ex4_loglistclass.LogDicts(logfilename)
-    result = ld.dicts()
+    result = ld.dicts(key=None)
```
Comment (Coraline13) [inline module4/ex4_loglistclass/ex4_loglistclass_test.py]: The tests should only be modified if it brings an improvement to the overall code. Here, for example, you should adapt the function to the tests by making `key` an optional argument with the default value as `None`.

File: module4/ex10_wedding_planner/ex10_wedding_planner_test.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def some_people():
 def populated_tables(some_people):
     gl = GuestList()
     for table_number, one_person in enumerate(some_people):
-        gl.assign(one_person, table_number)
+        gl.assign(one_person, table_number+1)
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner_test.py]: Here you should adapt your code instead of changing the test.

File: module4/ex4_loglistclass/ex4_loglistclass_test.py [hunk 3]
```diff
@@ -6,15 +6,15 @@
...
     assert type(result) == list
     assert all([type(one_item) == dict
                 for one_item in result])
 
 
 def test_iterdicts_returns_iterator_dicts():
     ld = ex4_loglistclass.LogDicts(logfilename)
-    result = ld.iterdicts()
+    result = ld.iterdicts(key=None)
```
Comment (Coraline13) [inline module4/ex4_loglistclass/ex4_loglistclass_test.py]: Same as above.

File: module4/ex4_loglistclass/ex4_loglistclass_test.py [hunk 4]
```diff
@@ -59,15 +59,15 @@ def test_latest():
 
 def test_for_ip():
     ld = ex4_loglistclass.LogDicts(logfilename)
-    matching_requests = ld.for_ip("65.55.106.183")
+    matching_requests = ld.for_ip("65.55.106.183", key=None)
```
Comment (Coraline13) [inline module4/ex4_loglistclass/ex4_loglistclass_test.py]: Same as above.

File: module4/ex4_loglistclass/ex4_loglistclass_test.py [hunk 5]
```diff
@@ -59,15 +59,15 @@ def test_latest():
...
     assert len(matching_requests) == 2
     assert all([one_item['ip_address'] == '65.55.106.183'
                 for one_item in matching_requests])
 
 
 def test_for_request():
     ld = ex4_loglistclass.LogDicts(logfilename)
-    matching_requests = ld.for_request("/robots.txt")
+    matching_requests = ld.for_request("/robots.txt", key=None)
```
Comment (Coraline13) [inline module4/ex4_loglistclass/ex4_loglistclass_test.py]: Same as above.

File: module4/ex10_wedding_planner/ex10_wedding_planner_test.py [hunk 6]
```diff
@@ -73,7 +73,7 @@ def test_table_free_space(some_people):
     for one_person in some_people[:GuestList.max_at_table - 1]:
         gl.assign(one_person, 1)
     assert (gl.free_space()) == {1: 1}
-
+    one_person = some_people[1]
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner_test.py:76]: Great catch! I will include this in the original repo.

File: module4/ex24_uniquish/ex24_uniquish.py [hunk 7]
```diff
@@ -2,16 +2,35 @@ class Foo:
...
+    def __hash__(self):
+         return hash(tuple(sorted(self.__dict__.items())))
 
 
 class Bar(Uniquish):
     def __init__(self, x, y):
         self.x = x
         self.y = y
 
+    def __eq__(self, other):
```
Comment (Coraline13) [inline module4/ex24_uniquish/ex24_uniquish.py]: You don't have to redefine `__eq__` and `__hash__` methods, that's what `Uniquish` class inheritance is for.

File: module4/ex17_formatted_cart/ex17_formatted_cart.py [hunk 8]
```diff
@@ -1,9 +1,30 @@
...
 class Item:
-    pass
+    def __init__(self, quantity, measure, name, price):
+        self.quantity = quantity
+        self.measure = measure
+        self.name = name
+        self.price = price
+
+    def __str__(self):
+        return f"{self.quantity:5} {self.measure:5} {self.name:10} @ ${float(self.price)}...$" \
```
Comment (Coraline13) [inline module4/ex17_formatted_cart/ex17_formatted_cart.py]: This is a good opportunity to make use of `try/except` and log an error message (if `price` is not a number or a string representing a number). You can then quit the program by throwing a `ValueError` exception.

File: module4/ex17_formatted_cart/ex17_formatted_cart.py [hunk 9]
```diff
@@ -1,9 +1,30 @@
...
+               f"{float(self.quantity*self.price)}"
```
Comment (Coraline13) [inline module4/ex17_formatted_cart/ex17_formatted_cart.py]: Also, it is best to only make the conversion of variable `price` to `float` once. You can do this in the `__init__` method.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 10]
```diff
@@ -1,10 +1,75 @@
...
+            self.add_to_table(person, table_number)
+
+    def table(self, table_number):
+        return self.tables[table_number]
+
+    def unassigned(self):
+        return self.tables['unassigned']
+
+    def free_space(self):
+        new_dict = {key: self.max_at_table - len(self.tables[key]) for key in self.tables}
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: You can directly return the value of `new_dict` if you are not using it for other purposes as well. It is also recommended not to use too generic variable names, but more specific ones to improve code readability.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 11]
```diff
@@ -1,10 +1,75 @@
...
 
+class GuestList:
+    max_at_table = 10
+
+    def __init__(self):
+        self.persons = {}
+        self.tables = {}
+
+    def __len__(self):
+        return len(self.persons.keys())
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: It is enough to just write `return len(self.persons)` because the length of a dictionary is equal to the number of key-value pairs it has.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 12]
```diff
@@ -1,10 +1,75 @@
...
+
+    def add_to_table(self, person, table_number):
+        if table_number:
+            if table_number in self.tables.keys():
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: It is enough to just write `if table_number in self.tables:` because `in` tests for the existence of a key in a dictionary.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 13]
```diff
@@ -1,10 +1,75 @@
...
+                self.tables[table_number].append(person)
+                if len(self.tables[table_number]) > self.max_at_table:
+                    raise TableFull("Table is already full")
+            else:
+                self.tables[table_number] = [person]
+        else:
+            if 'unassigned' in self.tables.keys():
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: Same as above.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 14]
```diff
@@ -1,10 +1,75 @@
...
+                self.tables['unassigned'].append(person)
+            else:
+                self.tables['unassigned'] = [person]
+
+    def assign(self, person, table_number):
+        if person in self.persons.keys():
```
Comment (Coraline13) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: Same as above.

File: module4/ex16_magic_tuples/ex16_magic_tuples.py [hunk 15]
```diff
@@ -1,5 +1,6 @@
-def magic_tuples():
-    pass
+def magic_tuples(total, max_value):
+    for i in range(total - max_value + 1, max_value):
```
Comment (Coraline13) [inline module4/ex16_magic_tuples/ex16_magic_tuples.py:2]: What if `max_value` is greater than `total`? For example, `magic_tuples(5, 8)`.

(I will make sure to add an extra test for this edge case.)

**Review by Coraline13 — CHANGES_REQUESTED** [review]

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 16]
```diff
@@ -1,10 +1,75 @@
...
+    max_at_table = 10
+
+    def __init__(self):
+        self.persons = {}
+        self.tables = {}
+
+    def __len__(self):
+        return len(self.persons.keys())
+
+    def add_to_table(self, person, table_number):
```
Comment (AurelGod) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: add_to_table is a private-like method so it should be renamed to _add_to_table.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 17]
```diff
@@ -1,10 +1,75 @@
...
+        if table_number:
+            if table_number in self.tables.keys():
+                self.tables[table_number].append(person)
+                if len(self.tables[table_number]) > self.max_at_table:
+                    raise TableFull("Table is already full")
```
Comment (AurelGod) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: I think this should be checked before appending the person.

File: module4/ex10_wedding_planner/ex10_wedding_planner.py [hunk 18]
```diff
@@ -1,10 +1,75 @@
...
+    def unassigned(self):
+        return self.tables['unassigned']
+
+    def free_space(self):
+        new_dict = {key: self.max_at_table - len(self.tables[key]) for key in self.tables}
+        return new_dict
+
+    def guests(self):
+        dict_tables = {key: sorted(self.tables[key], key=lambda person: (person.last, person.first))
+                       for key in self.tables if type(key) == int}
```
Comment (AurelGod) [inline module4/ex10_wedding_planner/ex10_wedding_planner.py]: Avoid to use generic variable names here, especially here where key is both dictionary key and key sort and it becomes a little bit confusing. For dictionary key something like table/table_nr as variable name could be used.

For sorting the list of persons, which are named tuples another option for sort key could be `key=attrgetter('last', 'first')`.

File: module4/ex14_thresholdequal/ex14_thresholdequal.py [hunk 19]
```diff
@@ -1,5 +1,29 @@
...
-    pass
+    def __init__(self, x):
+        self.x = x
+        self.tolerance = 2
+
+    def __eq__(self, other):
+        return abs(self.x - other.x) <= self.tolerance
+
+    def __add__(self, other):
+        return self.x + other.x
```
Comment (AurelGod) [inline module4/ex14_thresholdequal/ex14_thresholdequal.py]: I consider this should have the return value the same as the operand types. If the operands are ThresholdEqual the result should have the same type.

Also not necessary but would have been nice to have both add and iadd implemented and to emphasise the difference between them in the implementation.

The above refers to all the calculation operators below.

File: module4/ex14_thresholdequal/ex14_thresholdequal.py [hunk 20]
```diff
@@ -1,5 +1,29 @@
 class ThresholdEqual:
-    pass
+    def __init__(self, x):
+        self.x = x
+        self.tolerance = 2
+
+    def __eq__(self, other):
+        return abs(self.x - other.x) <= self.tolerance
```
Comment (AurelGod) [inline module4/ex14_thresholdequal/ex14_thresholdequal.py]: Please implement also lt and gt magic methods.

File: module4/ex4_loglistclass/ex4_loglistclass.py [hunk 21]
```diff
@@ -1,5 +1,49 @@
...
+        self.list_of_dicts = []
+        with open(filename, 'r') as f:
+            for line in f:
+                my_dict = {'ip_address': line.split(' ', 1)[0],
+                           'timestamp': line.split('[')[1].split(']')[0],
+                           'request': line.split('"')[1].split('"')[0]
+                           }
+                self.list_of_dicts.append(my_dict)
+
+    def dicts(self, key):
```
Comment (AurelGod) [inline module4/ex4_loglistclass/ex4_loglistclass.py]: Default parameter for key should be None, the same for other methods.

File: module4/ex4_loglistclass/ex4_loglistclass.py [hunk 22]
```diff
@@ -1,5 +1,49 @@
...
+            return new_list
+        else:
+            return self.list_of_dicts
+
+    def iterdicts(self, key):
+        new_list = self.list_of_dicts
+        if key:
+            new_list = sorted(self.list_of_dicts, key=key)
+        for one_dict in new_list:
+            yield one_dict
```
Comment (AurelGod) [inline module4/ex4_loglistclass/ex4_loglistclass.py:27]: Good job. Another way by using iter and dicts: return iter(self.dicts(key)).

File: module4/ex4_loglistclass/ex4_loglistclass.py [hunk 23]
```diff
@@ -1,5 +1,49 @@
...
+
+    def earliest(self):
+        ordered_list = sorted(self.list_of_dicts, key=lambda d: d['timestamp'])
```
Comment (AurelGod) [inline module4/ex4_loglistclass/ex4_loglistclass.py]: key should be a time object for an accurate comparison, please use time.strptime as stated in the exercise. Also this could be written in one line with return.

File: module4/ex4_loglistclass/ex4_loglistclass.py [hunk 24]
```diff
@@ -1,5 +1,49 @@
...
+        return ordered_list[0]
+
+    def latest(self):
+        ordered_list = sorted(self.list_of_dicts, key=lambda d: d['timestamp'])
+        return ordered_list[-1]
```
Comment (AurelGod) [inline module4/ex4_loglistclass/ex4_loglistclass.py]: Same as for earliest.

**Review by AurelGod — CHANGES_REQUESTED** [review]

File: module4/ex24_uniquish/ex24_uniquish.py [hunk 25]
```diff
@@ -2,9 +2,23 @@ class Foo:
...
 
 class Uniquish:
-    pass
+    def __eq__(self, other):
+        if type(self) == type(other):
+            return self.__dict__ == other.__dict__
+
+    def __hash__(self):
+        my_str_list = [str(item) for item in self.__dict__.values()]
+        return hash(tuple(my_str_list))
```
Comment (AurelGod) [inline module4/ex24_uniquish/ex24_uniquish.py]: Good for converting to string so to make sure the item is hashable.
Nothing to change here, I would've include also attribute name, value type along with the value to make sure about  a more in-depth equality.
Something like this would work:
hash(tuple((attr, str(value), type(value)) for attr, value in self.__dict__ .items()))

File: module4/ex14_thresholdequal/ex14_thresholdequal.py (hunk 19 above)
Comment (ralucaromascu) [reply module4/ex14_thresholdequal/ex14_thresholdequal.py]: If a function for operators would return a ThresholdEqual object instead of an int value, the  test_simple_int_stuff() test would no longer work.

File: module4/ex17_formatted_cart/ex17_formatted_cart.py [hunk 26]
```diff
@@ -1,9 +1,33 @@
...
+            raise TypeError("unsupported format string passed to Cart.__format__")
 
 
 class Item:
-    pass
+    def __init__(self, quantity, measure, name, price):
+        self.quantity = quantity
+        self.measure = measure
+        self.name = name
+        if type(price) == int:
```
Comment (Coraline13) [inline module4/ex17_formatted_cart/ex17_formatted_cart.py]: This only tests if `price` is of type `int`. But what if it's already of type `float`? Or a string representing a number? These cases would result in a `ValueError` being thrown.

I think the best approach here is a `try` - `except ValueError` block. This way you can catch any exception from trying to make the conversion `float(price)`.

Comment (AurelGod) [reply module4/ex17_formatted_cart/ex17_formatted_cart.py]: There a couple of things to consider here:
1. Constructors usually raise ValueError or TypeError for constructor arguments validation.
2. Depending on assumptions made one of these errors could be raised: 

- if allowed types for price are only strictly numeric values than it's enough to check price data type to be int or float and to raise a TypeError otherwise.
- if allowed types are numeric-like values (i.e. including strings representing a number) than a conversion to float could be done within a try-except block with ValueError.

So here you could approach the solution both ways depending on the assumptions made.

Another general thing to mention is the approach to validation, there are 2 main ways:
- EAFP - "It's easier to ask for forgiveness than permission" - this coding style assumes the existence of valid keys or attributes and catches exceptions if the assumption proves false
- LBYL - "Look before you leap" - this one prefers if statements over try-except blocks.
Pythons encourages EAFP style, though sometimes LBYL suits more.

File: module4/ex14_thresholdequal/ex14_thresholdequal.py [hunk 27]
```diff
@@ -1,5 +1,38 @@
...
+
+    def __gt__(self, other):
+        return self.x - other.x > 0 and abs(self.x - other.x) > self.tolerance
+
+    def __add__(self, other):
+        return self.x + other.x
+
+    def __iadd__(self, other):
+        self.x += other.x
+        return self
```
Comment (AurelGod) [inline module4/ex14_thresholdequal/ex14_thresholdequal.py]: Nothing to change here, just some mentions. 
- Saw indeed that tests do not pass if returning type for `__add__` is not a number but a new Threshold class. Despite this, returning a result of the same type as the operands themselves is the way operators should work. Here or the tests should be changed to consider this or for tests to have passed it should have to rewrite` __eq__`  so it compares Threshold objects also to numbers, but that would imply for consistency to implement the other operators with `other` being a number and it would have been too complicated.
- Nice that `__iadd__` operator is implemented as an in-place operation and return the object itself not a new one.
- For validation it could be validated that `x` is a number in init and `other` is a Threshold object (or a number).
